### PR TITLE
Remove DX12 descriptor heap compaction as well as disable descriptor allocation if unbounded arrays are not supported

### PR DESCRIPTION
**Pull request by moudgils**
## What does this PR do?

- With the addition of Bindless there is no need to compact descriptor heap anymore. If a use tries to use unbounded array in a custom SRG and fragments the descriptor heap I had added messages to encourage them to use Bindless Srg as that does not fragment.
- Disabled any bindless related descriptor allocation cost if unbounded arrays are not supported for Vulkan and Metal

## How was this PR tested?

Tested ASV as well as large levels within o3de-workshops-2022

### CODE REVIEW

File: Gems/Atom/RHI/DX12/Code/Source/RHI/DescriptorContext.cpp [hunk 1]
```diff
@@ -197,6 +178,14 @@ namespace AZ
             if (shaderResourceView.IsNull())
             {
                 shaderResourceView = AllocateHandle(D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV, D3D12_DESCRIPTOR_HEAP_FLAG_NONE, 1);
+                if (shaderResourceView.IsNull())
+                {
+                    AZ_Assert(
+                        false,
+                        "Descriptor heap ran out of memory for descriptor handles. Please consider increasing the first value of "
+                        "DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV within platformlimits.azasset file for dx12.");
```
Comment (spham-amzn) [inline Gems/Atom/RHI/DX12/Code/Source/RHI/DescriptorContext.cpp:186]: Should `DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV` be `D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV` instead? (and for the rest of the allocation asserts below)

File: Gems/Atom/RHI/DX12/Code/Source/RHI/ShaderResourceGroupPool.cpp [hunk 2]
```diff
@@ -109,7 +109,7 @@ namespace AZ
             }
             Base::ShutdownInternal();
         }
-
+
```
Comment (AMZN-alexpete) [inline Gems/Atom/RHI/DX12/Code/Source/RHI/ShaderResourceGroupPool.cpp:112]: nit, looks like an indent snuck in

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/Device.cpp [hunk 3]
```diff
@@ -1087,8 +1087,12 @@ namespace AZ
             // to determine if ray tracing is supported on this device
             StringList deviceExtensions = physicalDevice.GetDeviceExtensionNames();
             StringList::iterator itRayTracingExtension = AZStd::find(deviceExtensions.begin(), deviceExtensions.end(), VK_KHR_RAY_TRACING_PIPELINE_EXTENSION_NAME);
-            m_features.m_rayTracing = (itRayTracingExtension != deviceExtensions.end());
             m_features.m_unboundedArrays = physicalDevice.GetPhysicalDeviceDescriptorIndexingFeatures().shaderStorageTexelBufferArrayNonUniformIndexing;
+            if (m_features.m_unboundedArrays)
+            {
+                // Ray tracing needs unbounded arrays to work
+                m_features.m_rayTracing = (itRayTracingExtension != deviceExtensions.end());
+            }
```
Comment (spham-amzn) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/Device.cpp:1095]: nit: I had to do a double-take reading this based on the comment. Maybe:
```suggestion
            // Ray tracing needs raytracing extensions and unbounded arrays to work
            m_features.m_rayTracing = m_features.m_unboundedArrays && (itRayTracingExtension != deviceExtensions.end());

```

**Review by spham-amzn — APPROVED** [review]

File: Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/DeviceFeatures.h [hunk 4]
```diff
@@ -76,6 +76,9 @@ namespace AZ
             //! Whether Unbounded Array support is available.
             bool m_unboundedArrays = false;
 
+            //! Whether Unbounded Array support is simulated for Bindless Srg.
+            bool m_simulateBindlessUA = false;
```
[the remark this answers is not in the document]
Comment (moudgils) [reply Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/DeviceFeatures.h:80]: Either m_unboundArrays or m_simulateBindlessUA can be supported but m_unboundArrays should be given precedence over m_simulateBindlessUA. In future I am hoping that  we can simply remove m_simulateBindlessUA once we add proper unbounded array support using Apple's shader converter.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/BindlessDescriptorPool.cpp [hunk 5]
```diff
@@ -21,6 +21,10 @@ namespace AZ::Vulkan
     void BindlessDescriptorPool::Init(Device& device, const AZ::RHI::BindlessSrgDescriptor& bindlessSrgDesc)
     {
         m_device = &device;
+        if (!device.GetFeatures().m_unboundedArrays)
+        {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (moudgils) [reply Gems/Atom/RHI/Vulkan/Code/Source/RHI/BindlessDescriptorPool.cpp]: I made some updates based on your suggestion.

**Pull request merged**